File: finder4j-backtrack/src/test/java/com/github/k0zka/finder4j/backtrack/LocalParallelTrackTest.java
package com.github.k0zka.finder4j.backtrack;

import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.ForkJoinTask;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.runners.MockitoJUnitRunner;
import org.mockito.stubbing.Answer;

@RunWith(MockitoJUnitRunner.class)
public class LocalParallelTrackTest {
	@Mock
	ForkJoinPool pool;
	@Mock
	State state;
	@Mock
	StepFactory<Step<State>, State> stepFactory;
	@Mock
	TerminationStrategy<State> terminationStrategy;
	@Mock
	SolutionListener<State, Step<State>> listener;
	
	LocalParallelTrack<State, Step<State>> localParallelTrack;
	
	@Before
	public void setup() {
		localParallelTrack = new LocalParallelTrack<>(pool);
	}
	
	@Test
	public void available() {
		Mockito.when(pool.getActiveThreadCount()).thenReturn(10);
		Mockito.when(pool.getParallelism()).thenReturn(16);
		Assert.assertTrue(localParallelTrack.available());
	}

	@Test
	public void notAvailable() {
		Mockito.when(pool.getActiveThreadCount()).thenReturn(16);
		Mockito.when(pool.getParallelism()).thenReturn(16);
		Assert.assertTrue(localParallelTrack.available());
	}

	@SuppressWarnings("unchecked")
	@Test
	public void start() {
		Mockito.when(pool.submit(Mockito.any(ForkJoinTask.class))).thenAnswer(new Answer<ForkJoinTask<?>>() {

			@Override
			public ForkJoinTask<?> answer(InvocationOnMock invocation)
					throws Throwable {
				return null;
			}
		});
		localParallelTrack.start(state, stepFactory, terminationStrategy, listener);
		
	}
	
}

File: finder4j-backtrack/src/test/kotlin/com/github/k0zka/finder4j/backtrack/LocalParallelTrackTest.kt
package com.github.k0zka.finder4j.backtrack

import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import java.util.concurrent.ForkJoinPool

class LocalParallelTrackTest {

	private var pool: ForkJoinPool = mock()

	private var state: State = mock()

	private var stepFactory: StepFactory<Step<State>, State> = mock()

	private var terminationStrategy: TerminationStrategy<State> = mock()

	private var listener: SolutionListener<State> = mock()

	private var localParallelTrack: LocalParallelTrack<State, Step<State>>? = null

	@Before
	fun setup() {
		localParallelTrack = LocalParallelTrack(pool, stepFactory, terminationStrategy, listener)
	}

	@Test
	fun available() {
		whenever(pool.activeThreadCount).thenReturn(10)
		whenever(pool.parallelism).thenReturn(16)
		Assert.assertTrue(localParallelTrack!!.available())
	}

	@Test
	fun notAvailable() {
		whenever(pool.activeThreadCount).thenReturn(16)
		whenever(pool.parallelism).thenReturn(16)
		Assert.assertTrue(localParallelTrack!!.available())
	}

	@Test
	fun start() {
		whenever(
				pool.submit(any())).thenAnswer { null }
		localParallelTrack!!.start(state)

	}

}
